### PR TITLE
GTEST: prepare the infra for sockaddr testing

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

File: test/gtest/common/test_helpers.cc [hunk 1]
```diff
@@ -80,6 +81,48 @@ void safe_usleep(double usec) {
     }
 }
 
+void print_ip(char *if_name, struct sockaddr *ifa_addr)
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: const* for both

File: test/gtest/common/test_helpers.cc [hunk 2]
```diff
@@ -80,6 +81,48 @@ void safe_usleep(double usec) {
...
+{
+    size_t ip_len = ucs_max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN);
+    char ip_str[ip_len];
+
+    UCS_TEST_MESSAGE << "Testing " << if_name << " with " <<
+                        ucs_sockaddr_str(ifa_addr, ip_str, ip_len);
+}
+
+bool is_iface_ipoib(struct ifaddrs *ifa) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: const*

File: test/gtest/common/test_helpers.cc (hunk 1 above)
Comment (yosefe) [reply test/gtest/common/test_helpers.cc]: also, better have a function which returns std::string with the address - would be more generic

File: test/gtest/common/test_helpers.h [hunk 3]
```diff
@@ -72,6 +76,25 @@ int test_time_multiplier();
...
+ * IPv4 address on it.
+ */
+bool is_iface_ipoib(struct ifaddrs *ifa);
+
+
+/**
+ * From the given list of available interfaces, set addr to the first IPoIB
+ * address present. If none is present, set addr to NULL.
+ */
+void set_ip(struct ifaddrs **ifaddr, const struct sockaddr** addr);
```
Comment (yosefe) [inline test/gtest/common/test_helpers.h]: the function name is not related at all to what the function does
it should be
`struct sockaddr* find_ipoib_addr(struct ifaddrs **ifaddrs);`

File: test/gtest/common/test_helpers.cc (hunk 1 above)
Comment (alinask) [reply test/gtest/common/test_helpers.cc]: you mean that would only call ucs_sockaddr_str()?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5062/ for details (Mellanox internal link).

File: test/gtest/common/test_helpers.cc (hunk 1 above)
Comment (yosefe) [reply test/gtest/common/test_helpers.cc]: yes

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5064/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5065/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5066/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5067/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5071/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5072/ for details (Mellanox internal link).

**Comment by alinask** [comment]
bot:ornl:retest

**Comment by alinask** [comment]
@MattBBaker Can you please post the ORNL failure?

**Comment by alinask** [comment]
@yosefe I applied the comments, please review.

File: src/ucs/sys/sys.c [hunk 4]
```diff
@@ -614,6 +614,25 @@ ucs_status_t ucs_netif_ioctl(const char *if_name, unsigned long request,
     return status;
 }
 
+int ucs_netif_check(const char *if_name)
```
Comment (yosefe) [inline src/ucs/sys/sys.c]: ucs_netif_is_active

File: test/gtest/common/test_helpers.cc [hunk 5]
```diff
@@ -80,6 +81,35 @@ void safe_usleep(double usec) {
...
 }
 
+std::string get_iface_ip(const struct sockaddr *ifa_addr) {
+    size_t ip_len = ucs_max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN);
+    char ip_str[ip_len];
+
+    return ucs_sockaddr_str(ifa_addr, ip_str, ip_len);
+}
+
+bool is_ip_on_iface(const struct sockaddr* ifa_addr) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: is_inet_addr

File: test/gtest/common/test_helpers.cc [hunk 6]
```diff
@@ -80,6 +81,35 @@ void safe_usleep(double usec) {
...
+    return ((ifa_addr->sa_family == AF_INET) ||
+            (ifa_addr->sa_family == AF_INET6));
+}
+
+bool is_iface_on_dev(const char *ifa_name) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: is_ib_netdev()

File: test/gtest/common/test_helpers.cc [hunk 7]
```diff
@@ -80,6 +81,35 @@ void safe_usleep(double usec) {
...
+    char *path;
+    DIR *dir;
+    bool ret = false;
+
+    path = (char*)calloc(1, PATH_MAX);
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: can use local var char path[PATH_MAX];

File: test/gtest/common/test_helpers.cc [hunk 8]
```diff
@@ -80,6 +81,35 @@ void safe_usleep(double usec) {
...
+    snprintf(path, PATH_MAX, "/sys/class/net/%s/device/infiniband", ifa_name);
+
+    dir = opendir(path);
+    if (dir != NULL) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: if (dir == NULL) { return false; } else { closedir(dir); return true; }

File: test/gtest/common/test_helpers.h [hunk 9]
```diff
@@ -12,13 +12,18 @@
 
 #include <ucs/sys/preprocessor.h>
 #include <ucs/sys/checker.h>
+#include <ifaddrs.h>
+#include <net/if_arp.h>
```
Comment (yosefe) [inline test/gtest/common/test_helpers.h]: why need all those includes in test_helpers.h? can move to test_helpers.cc?

File: test/gtest/uct/uct_test.cc [hunk 10]
```diff
@@ -80,14 +89,43 @@ std::vector<const resource*> uct_test::enum_resources(const std::string& tl_name
...
                 rsc.dev_type   = tl_resources[j].dev_type;
                 all_resources.push_back(rsc);
             }
 
+            if (md_attr.cap.flags & UCT_MD_FLAG_SOCKADDR) {
+                struct ifaddrs *ifaddr, *ifa;
+                ucs_sock_addr_t sock_addr;
+
+                EXPECT_TRUE(getifaddrs(&ifaddr) != -1);
+                for (ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next) {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: need to extract code to separate function

File: test/gtest/uct/uct_test.cc [hunk 11]
```diff
@@ -80,14 +89,43 @@ std::vector<const resource*> uct_test::enum_resources(const std::string& tl_name
...
+                    sock_addr.addr = ifa->ifa_addr;
+                    if (uct_md_is_sockaddr_accessible(pd, &sock_addr, UCT_SOCKADDR_ACC_REMOTE) &&
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: what about ACC_LOCAL?

File: test/gtest/uct/uct_test.cc [hunk 12]
```diff
@@ -80,14 +89,43 @@ std::vector<const resource*> uct_test::enum_resources(const std::string& tl_name
...
+                       (ucs_netif_check(ifa->ifa_name))) {
+                        resource rsc;
+                        rsc.md_name    = md_resources[i].md_name;
+                        rsc.local_cpus = md_attr.local_cpus;
+                        rsc.tl_name    = "sockaddr";
+                        rsc.dev_name   = ifa->ifa_name;
+                        rsc.dev_type   = UCT_DEVICE_TYPE_LAST;
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: why?

File: test/gtest/uct/uct_test.cc [hunk 13]
```diff
@@ -80,14 +89,43 @@ std::vector<const resource*> uct_test::enum_resources(const std::string& tl_name
...
+
+                        if (ifa->ifa_addr->sa_family == AF_INET) {
+                            *(struct sockaddr_in*)&(rsc.if_addr)  = *(struct sockaddr_in*) ifa->ifa_addr;
+                        } else if (ifa->ifa_addr->sa_family == AF_INET6) {
+                            *(struct sockaddr_in6*)&(rsc.if_addr) = *(struct sockaddr_in6*) ifa->ifa_addr;
+                        } else {
+                            UCS_TEST_ABORT("Unknown sa_family" << ifa->ifa_addr->sa_family);
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: add space after "sa_family"

File: test/gtest/uct/uct_test.cc [hunk 14]
```diff
@@ -252,8 +290,11 @@ uct_test::entity::entity(const resource& resource, uct_iface_config_t *iface_con
 
     ucs_status_t status;
 
-    params->mode.device.tl_name    = const_cast<char*>(resource.tl_name.c_str());
-    params->mode.device.dev_name   = const_cast<char*>(resource.dev_name.c_str());
+    if (params->open_mode == UCT_IFACE_OPEN_MODE_DEVICE) {
+        params->mode.device.tl_name    = const_cast<char*>(resource.tl_name.c_str());
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: const_cast should not be needed here

File: test/gtest/uct/uct_test.cc (hunk 11 above)
Comment (alinask) [reply test/gtest/uct/uct_test.cc]: will add so that both will be checked.

File: test/gtest/uct/uct_test.cc (hunk 12 above)
Comment (alinask) [reply test/gtest/uct/uct_test.cc]: if this field isn't set, valgrind complains about it when checking it in several tests.
for example - https://github.com/openucx/ucx/blob/master/test/gtest/uct/uct_p2p_test.cc#L41

Comment (yosefe) [reply test/gtest/uct/uct_test.cc]: should be NET

**Comment by MattBBaker** [comment]
C++ error. in uct_test.cc
```
  CXX    uct/gtest-test_tag.o
cc1plus: warnings being treated as errors
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc: In static member function ‘static std::vector<const resource*, std::allocator<const resource*> > uct_test::enum_resources(const std::string&, bool)’:
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:117: error: dereferencing pointer ‘<anonymous>’ does break strict-aliasing rules
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:117: note: initialized from here
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:119: error: dereferencing pointer ‘<anonymous>’ does break strict-aliasing rules
/var/lib/jenkins/jobs/UCX/workspace/contrib/../test/gtest/uct/uct_test.cc:119: note: initialized from here
make[2]: *** [uct/gtest-uct_test.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/1510072926/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/1510072926'
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
```

File: test/gtest/common/test_helpers.h (hunk 9 above)
Comment (alinask) [reply test/gtest/common/test_helpers.h]: only #include <netinet/in.h> is needed for sockaddr which is also used in the header file. the rest are leftovers - i'll remove.

**Comment by alinask** [comment]
@MattBBaker Thanks! what is the gcc version you have on the jenkins host?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5118/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
gcc version is 'Red Hat 4.4.7-16'

**Comment by alinask** [comment]
thanks @MattBBaker , I updated the pod9 wiki page with this version.
I have a host with 'Red Hat 4.4.7-18' on which this error didn't reproduce.

**Comment by alinask** [comment]
@yosefe Can you please review?

**Comment by MattBBaker** [comment]
The latest merge also has it passing on pod9, so it looks like something in the update resolved this.

**Comment by alinask** [comment]
I changed the code a bit to overcome this error from gcc.

File: test/gtest/common/test_helpers.h (hunk 9 above)
Comment (yosefe) [reply test/gtest/common/test_helpers.h]: actually, struct sockaddr is defined in sys/socket.h

File: test/gtest/uct/uct_test.cc [hunk 15]
```diff
@@ -55,6 +56,41 @@ uct_test::~uct_test() {
     uct_config_release(m_md_config);
 }
 
+void uct_test::set_sockaddr_resources(uct_md_h pd, char *md_name, cpu_set_t local_cpus,
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc]: minor: s/pd/md

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5137/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe Can this be merged?

**Pull request merged**